### ISSUE TITLE
Debugging webpacked extension debugs stale version of code

### ISSUE THREAD

**Issue by itowlson** [microsoft/vscode#82076]
Issue Type: <b>Bug</b>

1. Clone https://github.com/itowlson/vscode-kubernetes-tools
2. Switch to the `webpack-2` branch
3. `npm install` and install extension dependencies
4. F5 to debug and create a `.yaml` file to activate the extension. Exit the debugger
5. This should all be fine so far
6. In `extension.ts`, after line 133, add the following line: `console.log('***** hello');`.  DO NOT SAVE THE FILE!
7. Press F5 to debug.  Expected: I should see `***** hello` in the debug console.  Actual: I don't see the new message
8. Exit the debugger.  Edit the line to read `console.log('***** goodbye');`
9. Save the file and IMMEDIATELY hit F5.  Expected: I should see `***** goodbye` in the debug console.  Actual: I now see `***** hello` from the previous build which was the last one to complete.
10. If you edit the message again and save the file _then wait for a slow count of 3 before debugging_, then you _do_ pick up the update

VS Code version: Code 1.38.1 (b37e54c98e1a74ba89e03073e5a3761284e3ffb0, 2019-09-11T13:35:15.005Z)
OS version: Windows_NT x64 10.0.18912

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Intel(R) Xeon(R) CPU E5-1650 v2 @ 3.50GHz (12 x 3492)|
|GPU Status|2d_canvas: enabled<br>flash_3d: enabled<br>flash_stage3d: enabled<br>flash_stage3d_baseline: enabled<br>gpu_compositing: enabled<br>multiple_raster_threads: enabled_on<br>native_gpu_memory_buffers: disabled_software<br>oop_rasterization: disabled_off<br>protected_video_decode: unavailable_off<br>rasterization: enabled<br>skia_deferred_display_list: disabled_off<br>skia_renderer: disabled_off<br>surface_synchronization: enabled_on<br>video_decode: enabled<br>viz_display_compositor: disabled_off<br>webgl: enabled<br>webgl2: enabled|
|Load (avg)|undefined|
|Memory (System)|63.93GB (34.29GB free)|
|Process Argv|--folder-uri file:///d%3A/GitHub/vscode-kubernetes-tools|
|Screen Reader|no|
|VM|0%|
</details><details><summary>Extensions (19)</summary>

Extension|Author (truncated)|Version
---|---|---
better-toml|bun|0.3.2
gitlens|eam|10.1.0
tsl-problem-matcher|eam|0.0.4
EditorConfig|Edi|0.13.0
vscode-pull-request-github|Git|0.11.2
terraform|mau|1.4.0
vscode-docker|ms-|0.8.1
duffle-vscode|ms-|0.3.0
python|ms-|2019.9.34911
remote-wsl|ms-|0.39.5
azure-account|ms-|0.8.4
csharp|ms-|1.21.4
Go|ms-|0.11.7
powershell|ms-|2019.9.0
vscode-typescript-tslint-plugin|ms-|1.2.2
quicktype|qui|12.0.46
java|red|0.50.0
vscode-yaml|red|0.5.3
rust|rus|0.6.3

</details>

**Comment by jrieken** [comment]
Maybe a task watcher issue or debug related... Starting with @alexr00

**Comment by alexr00** [comment]
I think this is expected. The last compile to complete is the one that you'll debug. @itowlson am I understanding the issue correctly? There's not much we can do if you hit debug before the compile is done.

**Comment by alexr00** [comment]
We do have an issue somewhere for stopping the prelauncTask when you stop debugging. 

I think I'm seeing the issue better now. You're saying that the task should have been considered "incomplete" by debugging and so the debug session should have waited to start until it "re-completed". I'll look into what's going on here.

**Comment by itowlson** [comment]
@alexr00 Yes, your second comment is right.  When I hit F5 a non-webpack configuration, TypeScript and VS Code make sure that TS (in its watch mode) has finished dealing with my most recent changes, so that I am always debugging the code I "F5"-ed on.  When I hit F5 in a webpack configuration, the debugger kicks off before webpack (in _its_ watch mode) has finished processing the changes, so I end up debugging stale code.  Hope that is clearer, and sorry my original explanation was confusing.

**Comment by alexr00** [comment]
Debug checks for "active" tasks before running. The watch task is running, which means it's active. Debug sees this and goes, even if the watch task is currently busy recompiling. 

@isidorn I can add more info to what is returned by `getActiveTasks` so that you know that the task is currently busy and you should wait for an inactive task event on it (meaning that it is no longer busy. The term active is overloaded here.). What do you think?

**Comment by isidorn** [comment]
@alexr00 adding more info so we cover this case makes sense.
Also all the debug interactions with task is done in [this method](https://github.com/Microsoft/vscode/blob/7e57821af24dc167b910bda38d5e14ceb7bbb0b4/src/vs/workbench/contrib/debug/browser/debugService.ts#L753) feel free to modify it as you think is best (as you have already done with the PR). Thanks!

**Comment by alexr00** [comment]
I have pushed a change that should improve this. @itowlson can you install the next insiders build (after 2019-10-11T05:44:05.506Z) and see if you can still see the issue? 

I don't think we can completely fix this, but I think this might make it so you rarely see it.

**Comment by itowlson** [comment]
@alexr00 With the latest insiders build and the project and branch listed above, when I hit F5 to debug, I now get a few seconds' wait and then the error message "The specified task cannot be tracked."  Is there any additional information I can gather to help with diagnosing this?

![image](https://user-images.githubusercontent.com/865538/66969058-8076f480-f0e4-11e9-8e73-a05dfc70637b.png)

**Comment by itowlson** [comment]
To be clear, I'm more than prepared to believe this is something _I'm_ doing wrong, as I assume other people are webpacking their extensions and it must be working for them!  I just can't work out _what_ I'm doing wrong, and I believe I'm following the guidance the team gave me in #79956.

**Comment by alexr00** [comment]
Looks like something changed with either the problem matcher or webpack. The end pattern provided by the `$ts-webpack-watch problem matcher` is `Compilation (.*?)finished`. But that never gets printed to the terminal. @eamodio do you know anything about this?

Similarly, the begins pattern `Compilation (.*?)starting…` also doesn't show up when a file changes.

**Comment by eamodio** [comment]
Hrm, webpack could have changed something here (the problem matchers are extremely fragile). @itowlson do you have `--info-verbosity verbose` specified on your watch command?

**Comment by itowlson** [comment]
@alexr00 @eamodio I didn't have `--info-verbosity verbose` specified on `watch`.  When I add that (and set the `reveal` to `always` in `tasks.json`), here is what I see.  All this is with Insiders version 1.40.0-insider (user setup) commit 6a2d035b7157cb135b51d2d3ca3b3c0ee0fc4ad6 (full version info pasted after task output report).  It looks to me like the webpack watch might be taking longer than VS Code expects and the error dialog is jumping in too early... but here's the raw info, you folks are the experts!  Hope it helps - thanks for your help!

**Make a change and hit F5 (without manual save)**

```
> Executing task: npm run watch <


> vscode-kubernetes-tools@1.0.4 watch D:\GitHub\vscode-kubernetes-tools
> webpack --mode none --watch --info-verbosity verbose


Compilation  starting…


webpack is watching the files…


Compilation  finished

Hash: d7de210b4577a8148cec
Version: webpack 4.35.2
Time: 18611ms
Built at: 2019-10-18 09:21:07
           Asset      Size  Chunks             Chunk Names
    extension.js  10.3 MiB       0  [emitted]  main
extension.js.map    12 MiB       0  [emitted]  main
Entrypoint main = extension.js extension.js.map
  [0] ./src/extension.ts 89.8 KiB {0} [built]
  ... files elided ...
[147] ./src/explainer.ts 13.3 KiB {0} [built]
    + 928 hidden modules
```

_Notes:_ The change I made before hitting F5 *is* picked up and debugging works as desired.  However, there is a very long wait (maybe 20 seconds?) for this initial build.

**Make another change and again hit F5 (without manual save**

```
Compilation  starting…
```

Then after a few seconds the "The specified task cannot be tracked" dialog appears.  And then a few seconds after _that_ the task terminal resumes with:

```
Compilation  finished

Hash: abe020f4173febd97b59
Version: webpack 4.35.2   
Time: 13747ms
Built at: 2019-10-18 09:22:34
           Asset      Size  Chunks             Chunk Names
    extension.js  10.3 MiB       0  [emitted]  main
extension.js.map    12 MiB       0  [emitted]  main
Entrypoint main = extension.js extension.js.map
  [0] ./src/extension.ts 89.8 KiB {0} [built]
  ... files elided ...
[147] ./src/explainer.ts 13.3 KiB {0} [built]
    + 928 hidden modules
```

In this case debug doesn't proceed because the dialog doesn't give me an option to 'debug anyway,' so I can't tell whether the change would have been picked up.

**Make another change and hit Save (without debugging)**

```
Compilation  starting…


Compilation  finished

Hash: 05a0921aa70ee5baf42f
Version: webpack 4.35.2
Time: 10663ms
Built at: 2019-10-18 09:24:09
           Asset      Size  Chunks             Chunk Names
    extension.js  10.3 MiB       0  [emitted]  main
extension.js.map    12 MiB       0  [emitted]  main
Entrypoint main = extension.js extension.js.map
  [0] ./src/extension.ts 89.8 KiB {0} [built]
  ... files elided ...
[147] ./src/explainer.ts 13.3 KiB {0} [built]
    + 928 hidden modules
```

_Notes:_ If I wait for this to complete and then hit F5 then the change is picked up. However, If I hit F5 shortly after saving (which is my normal real-world workflow), then I get the tracking error dialog again.

**VS Code version information**

Version: 1.40.0-insider (user setup)
Commit: 6a2d035b7157cb135b51d2d3ca3b3c0ee0fc4ad6
Date: 2019-10-17T05:28:00.123Z
Electron: 6.0.12
Chrome: 76.0.3809.146
Node.js: 12.4.0
V8: 7.6.303.31-electron.0
OS: Windows_NT x64 10.0.18912

**Comment by eamodio** [comment]
@itowlson I tried this out and I can't reproduce the case you mention where you get an error dialog. But for the case where you hit F5 quickly after saving, I can reproduce that debugging will start immediately. And unfortunately I believe this is behaving as designed.

From what I understand, what is basically happening is that the problem matcher is looking to see if there is an ending pattern after a starting pattern (starting from the most recent output in the terminal). Which if you save and hit F5 too quickly is the case, because webpack hasn't yet written out `Compilation  starting…` to the terminal yet -- so it sees the previous starting and ending pair so everything looks good and it starts. If you wait to hit F5 until you see `Compilation  starting…` in the terminal for the watch task then it will wait until `Compilation  finished` is output.

I don't think there is much vscode can do here, because it is just a timing issue between the editor save, webpacks file watcher picking it up and starting a compile, and how quickly you hit F5.

As for the delay, that just seems to be the time it takes webpack to compile. Although I did notice you are passing `--mode none` to webpack watch, and if you use `--mode development` it seems to double the speed of your recompiles.

Hope that helps!

**Comment by itowlson** [comment]
Thanks @eamodio for the investigation.  Your advice definitely helps and I appreciate you taking the time to write it up.  You are right that `--mode development` makes a _huge_ difference.  At the moment that looks like it cuts the time enough that I can `F5` or `Ctrl+S + F5` without getting the tracking dialog, which I think must be some kind of timeout.  This may be brittle if compilation times go up again, but between Alex's fix and that flag I think we have a way forward at least for now.

Thanks to both of you for your time and diligence!

**Comment by eamodio** [comment]
@itowlson no problem. FYI, you can probably reduce your compile time and extension size by switching to compile to `esnext` modules (rather than `commonjs`) in your `tsconfig.json` (see [here](https://github.com/eamodio/vscode-gitlens/blob/59d9071a9bb200c3ef95369a6335f8a0138f461a/tsconfig.json)) and then switching some dependencies like `lodash` to use `lodash-es`. Those will take advantage of tree shaking.

Also I doubt these will help much, but you can move all the `@types/*` to `devDependencies` since those are only needed for compiling. You can remove the `clipboardy` and `opn` dependencies by utilizing the built-in vscode apis for both (on `vscode.env`). Also depending on what you are using `moment` for it is a LARGE dependency with all its languages -- so you could exclude those languages (if unneeded from your webpack config) or switch to sometime far more lightweight like `day.js` or `date-fns` (I use day.js in GitLens because it was smaller and faster for my needs).

Anyway, hope that helps.

**Comment by eamodio** [comment]
@alexr00 Do you know of a timeout on the task where we would give up?

**Comment by alexr00** [comment]
@eamodio for background tasks used with a launch config we just keep waiting once we see the begins pattern. Adding a timeout to `"background"` in the problem matcher could be a reasonable feature though.

**Comment by roblourens** [comment]
Not sure what to verify here - even if I wait until the task has printed "Compilation starting..." then press F5, the debug session starts immediately and runs the old code. It sounds like now it is supposed to wait for the background task to be complete but I am not sure.

**Comment by itowlson** [comment]
@roblourens Yes - if the user hits F5 then it should debug the code _as it exists when F5 was hit_, which means it needs for any background compilation to catch up.  Sounds like a lot of us are seeing slightly different behaviours though!  I provided my steps at the top of the issue though I think I might need to give you a specific commit because I pushed some of Eric's fixes to that branch, so it may no longer exhibit the behaviour I originally described - let me know if you want me to provide a specific commit ID.

**Comment by eamodio** [comment]
@roblourens are you using the TypeScript Webpack problem machers package?

**Comment by alexr00** [comment]
Looks like I have a bug where it waits for the busy task.

**Comment by alexr00** [comment]
Verification steps (slightly modified from the original issue steps):
1. Clone https://github.com/itowlson/vscode-kubernetes-tools
2. Switch to the `webpack-2` branch
3. In the package.json, change the `watch` script to be `"watch": "webpack --watch --info-verbosity verbose",`  This makes if easier to repro.
3. `npm install` and install extension dependencies
4. F5 to debug and create a `.yaml` file to activate the extension. Exit the debugger
5. This should all be fine so far
6. This bit involves careful timing:
- Open the terminal running your watch task (should still be going from your debug session
- Add a console.log in the extension activation
- Save your console.log **and wait until you see the "Compilation  starting…" message in the task terminal**
- **As soon as you see that message** hit F5 to debug
7. You should see the debugger wait until the compilation is finished before launching
8. When the extension is activated, you should see your log, indicating that it is correctly using the newly compiled code.

**Comment by chrmarti** [comment]
It waits, but then kind of times out with:
![image](https://user-images.githubusercontent.com/9205389/67952302-66add500-fbed-11e9-82e3-d6f622687a5b.png)

**Comment by alexr00** [comment]
@chrmarti were you testing out of source? It isn't in insiders yet.

**Comment by connor4312** [comment]
Running out of master, looks like it works!
